### DISCUSSION TITLE
[Wrong Usage] Server Routing and Client Routing both loaded.

### DISCUSSION THREAD

**Discussion by brillout** [vikejs/vike#750] [Discussion] [answered]
Moved from Discord:

> Hi, I'd appreciate some help/advice with this problem. I've upgraded vite-plugin-ssr to 0.4.108 and now I'm getting `Uncaught Error: [vite-plugin-ssr@0.4.108][Wrong Usage] Server Routing and Client Routing both loaded. Make sure your client-side code doesn't load both at the same time.` but I have no idea why. I have a feeling the bundler includes both routing systems because I import navigate in some of my React components. It has to do this because the components are isomorphic and the navigate only fires in a useEffect hook under certain conditions.

**Comment by brillout** [comment, marked as answer]
What's the stack trace the error?

**Reply by robinelvin** [reply]
```
entry-client-routing.2b48210e.js:233 Uncaught Error: [vite-plugin-ssr@0.4.108][Wrong Usage] Server Routing and Client Routing both loaded. Make sure your client-side code doesn't load both at the same time.
    at _j (entry-client-routing.2b48210e.js:234:828)
    at entry-client-routing.2b48210e.js:247:23031
$g @ entry-client-routing.2b48210e.js:233
re @ entry-client-routing.2b48210e.js:234
_j @ entry-client-routing.2b48210e.js:234
(anonymous) @ entry-client-routing.2b48210e.js:247
```
This only happens in production mode therefore the code is minified

**Reply by brillout** [reply]
Can you upload `entry-client-routing.2b48210e.js` e.g. to a secret GH gist? Or, if you prefer, reproduction welcome.

**Reply by robinelvin** [reply]
Gist: https://gist.github.com/robinelvin/fd08620636e52c9133d1d0156d8f66b9

Slightly different hash as I added some console logging to try to work out what's happening. Can provide a clean version if you prefer.

Thanks

**Reply by brillout** [reply]
Actually, I'm seeing that the stack trace is missing some column numbers, so it isn't helpful.

I'll need a reproduction.

Even if you fix the problem, a reproduction will be helpful as it will allow me to improve the error message. Alternatively, you can try to disable minification so we get a helpful stack trace. Keep me updated.

**Reply by robinelvin** [reply]
Ok, I'll see if I can get a reproduction working.
So far I have traced it back to the call to `onClientEntry_ServerRouting` happening in the browser - I don't understand the reason behind this but I'll keep digging.

**Reply by robinelvin** [reply]
I've tried recreating this issue in a blank project without success. I've spent several days on this so far and I'm still no further forward understanding the problem. I don't understand how, if I have enabled client routing, the client thinks server routing is also enabled. Perhaps if you can explain how this might happen or how to trace this I can investigate further. Thanks

**Reply by brillout** [reply]
Have a look at how [this file](https://github.com/brillout/vite-plugin-ssr/blob/617f7d497d0b8c1d5cb19a2c12b7a2f6562c124e/vite-plugin-ssr/utils/assertSingleInstance.ts) works. Inspect why the assertion is failing by injecting a couple of `console.log(new Error().stack)` directly at `node_modules/vite-plugin-ssr/dist/`. Also make sure to `$ rm -r node_modules/.vite/`.

**Reply by robinelvin** [reply]
I get how that file works. The stack trace isn't much use as both the calls are coming from directly imported inline Javascript. The first one is for client routing and then the next one tries to enable server routing. These files appear to be pulled in as part of the build but I don't know why. This problem does not happen when using `vite serve`

![image](https://user-images.githubusercontent.com/1187032/229770278-d4486a0d-9c6f-4597-80fe-a2a75f829a0a.png)

![image](https://user-images.githubusercontent.com/1187032/229770145-068f2ec3-8157-4d79-acfd-297e1dbcaf55.png)

**Reply by brillout** [reply]
Do you use Client Routing *everywhere*? Is there a page that uses Server Routing? If you remove all *other* `/pages/*` do you still get the error?

**Reply by robinelvin** [reply]
Success!

I wasn't using client routing everywhere as some pages need to be done server-side. On your suggestion I removed all pages except one and I still got the problem. Then, I don't know why I didn't think of it before, I looked at my Vite config and noticed a difference between your example and my setup. I removed `splitVendorChunkPlugin` and everything started working as expected.

TL;DR `splitVendorChunkPlugin` breaks the build - confirmed by adding to the `https://github.com/brillout/vite-plugin-ssr/tree/main/boilerplates/boilerplate-react-ts` project

**Reply by brillout** [reply]
I see, makes sense. Thanks for reporting back.

Btw., in case you think you can convince your employer: https://github.com/sponsors/brillout. Sponsors are starting to make a tremendous impact.

**Reply by brillout** [reply]
FYI I improved the error message https://github.com/brillout/vite-plugin-ssr/commit/caab024f42fb29cd3a4693b71b96f46135672f8e.

> The client runtime of Server Routing and the client runtime of Client Routing are both being loaded. Make sure your client-side code doesn't include(/bundle) both for a given page.

And made it a warning instead of an error (there is chance the app can still function).

**Comment by ser1us** [comment]
This caused a big headache for me today. The client router took over and tried to do execute some server route functions...

What I had was this entry in the `vite.config.ts` file:

```
output: {
  manualChunks(id) {
    if (id.includes("node_modules")) {
      ....
    }
  }
}
```

Basically I wanted to split out the node_module stuff into different chunks in order to keep the important parts small. but that completely messed with `vite-plugin-ssr`. I was wondering if I could help out with some documentation for this, like a heads up? Tough I do not know where to put it 🤔 Or even better since the splitting seems to be the issue, should we adjust the warning once more and include this information?

Cheers!
Christoph

**Reply by brillout** [reply]
Maybe we can add:

> Did you configure Rollup to include both routing in the same bundle (which we recommend against)?

Would that have helped you?

**Reply by brillout** [reply]
I went ahead with https://github.com/brillout/vite-plugin-ssr/commit/62a4bf848a20101e14392207f3599040fecbc463. Let me know if you believe the error message can still be improved.

**Reply by ser1us** [reply]
Fantastic, thank you very much! 

Just one minor thing, maybe change the wording from
`Did you configure Rollup to include both in a same bundle?`

to :
`Did you configure rollup to do chunk splitting?`

just so the terms align a little bit better with the actual settings.

**Reply by brillout** [reply]
This should do it: https://github.com/brillout/vite-plugin-ssr/commit/ccf385dec4788cc19e0e66e75a76fff464f502bc.

Let me know if you still think there is room for improvement. Or if there is anything else.

Looking forward to further polish VPS for the [v1 release](https://github.com/brillout/vite-plugin-ssr/labels/v1%20release%20%3Astar2%3A).

**Reply by ser1us** [reply]
That's the gold version :)

**Reply by brillout** [reply]
> gold

Speaking of which 😃 in case you're able to convince your employer: https://github.com/sponsors/brillout &mdash; it makes a massive difference. (I know you did some personal sponsoring already and thanks for that 💚.)
